src/mbr.c

```diff
@@ -0,0 +1,438 @@
+#include <unistd.h>
+#include <sys/ioctl.h>
+#include <sys/mount.h>
+#include <linux/hdreg.h>
+
+#include "mbr.h"
+#include "update_handler.h"
+
+/* partition entry in MBR partition table, the system boots from */
+#define BOOT_PARTITION_ENTRY		0
+#define MBR_NUMBER_OF_PARTITIONS	4
+#define MBR_MAGIC_NUMBER_L		0x55
+#define MBR_MAGIC_NUMBER_H		0xAA
+
+struct chs_entry {
+	guint8 head;
+	guint8 sector;
+	guint8 cylinder;
+};
+
+struct mbr {
+	guint8 bootstrap_code[440];
+	guint32 disk_signature;
+	guint16 unused;
+	struct partition_tbl_entry {
+		guint8 boot_indicator;
+		struct chs_entry chs_start;
+		guint8 type;
+		struct chs_entry chs_end;
+		guint32 partition_start;
+		guint32 partition_size;
+	} partition_table[MBR_NUMBER_OF_PARTITIONS];
+	guint8 magic_number[2];
+} __attribute__((packed));
+
+G_STATIC_ASSERT(sizeof(struct mbr) == 512);
+
+static guint get_sectorsize(gint fd)
+{
+	guint sector_size;
+
+	if (ioctl(fd, BLKSSZGET, &sector_size) != 0)
+		return 512;
+
+	return sector_size;
+}
+
+static gboolean get_number_of_sectors(gint fd, guint *sectors,
+		GError **error)
+{
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+	g_assert_nonnull(sectors);
+
+	if (ioctl(fd, BLKGETSIZE, sectors) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"ioctl command 0x%04x failed: %s",
+				BLKGETSIZE, g_strerror(errno));
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean get_hd_geometry(gint fd, struct hd_geometry *geometry,
+		GError **error)
+{
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+	g_assert_nonnull(geometry);
+
+	if (ioctl(fd, HDIO_GETGEO, geometry) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"ioctl command 0x%04x failed: %s",
+				HDIO_GETGEO, g_strerror(errno));
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean validate_region(gint fd, guint64 start, guint64 size,
+		guint sector_size, GError **error)
+{
+	gboolean res = FALSE;
+	guint number_of_sectors;
+	GError *ierror = NULL;
+
+	g_return_val_if_fail(error == NULL || *error == NULL, FALSE);
+
+	if (start < sizeof(struct mbr) || size == 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"no valid configuration for region");
+		goto out;
+	}
+
+	if ((start % (2 * sector_size)) != 0 || (size % (2 * sector_size)) != 0) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"region configuration is aligned to the double"
+				"sector-size %d", 2 * sector_size);
+		goto out;
+	}
+
+	res = get_number_of_sectors(fd, &number_of_sectors, &ierror);
+	if (!res) {
+		g_propagate_error(error, ierror);
+		goto out;
+	}
+
+	if ((start + size) >= (guint64)number_of_sectors * sector_size) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"region configuration is bigger than device");
+		res = FALSE;
+		goto out;
+	}
+
+out:
+	return res;
+}
+
+static gboolean read_mbr(gint fd, struct mbr *mbr, GError **error)
+{
+	gboolean res = FALSE;
```

Comment:
No need to have `res` variable here.

Simplify structure by instantly returning with `return FALSE` in if clauses and do `return TRUE` at the end of the function.